### PR TITLE
#166238094 Users should receive a JWT on Successful Registration/Login

### PR DESCRIPTION
**Pull request by ayodejiAA**
#### What does this PR do?
- [x] Send token to users upon successful registration or login

#### Description of Task to be completed?
- [x] Have authentication working on the following routes- 
```/api/v1/auth/signup``` and ```/api/v1/auth/login```

#### How should this be manually tested?
This feature can be tested with  any API testing application, like postman. It can as well be tested on the swagger documentation route. 
- Clone the REPO and checkout to ft-implement-jwt-authentication-166238094. Add all required env variables. 
- Run ```npm run buildTables``` follow by ```npm start```, and use your API testing app to test the above API endpoints. 
>- Required fields for signup are ```userName, email, firstName, lastName and password```
>- Required fields for login are ```userLogin (accepts email/userName) and password```

You can also point your browser to ```/api-docs/``` route to test the feature on swagger API documentation. 

#### Any background context you want to provide?
Authentication has always been an important aspect of every application. During registration and login, user should be provided token that will be required for authentication and use on subsequent calls to other application endpoints. 

#### What are the relevant pivotal tracker stories?
[#166238094](https://www.pivotaltracker.com/story/show/166238094)

#### Screenshots (if appropriate)
---
sam1

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "Book library management system",
   "main": "index.js",
   "scripts": {
-    "test": "cross-env NODE_ENV=test nyc --reporter=html --reporter=text mocha ./server/test --require @babel/register --colors --slow 400 --timeout 5000 --exit",
+    "test": "npm run clean-tables && cross-env NODE_ENV=test nyc --reporter=text mocha ./server/test/*.spec.js --require @babel/register --require @babel/polyfill --slow 400 --timeout 5000 --exit",
```
Comment (abdulfatai360) [inline package.json]: Bro, well done for adding the `babel polyfill` plugin to the test script. However, if I may ask, why did you remove the `--colors` option?

File: package.json [hunk 2]
```diff
@@ -28,8 +29,10 @@
   "homepage": "https://github.com/los-deferal-02/Hello-Books#readme",
   "dependencies": {
     "@babel/polyfill": "^7.4.4",
+    "bcrypt": "^3.0.6",
```
Comment (abdulfatai360) [inline package.json]: Bro, sorry. Bcrypt doesn't work on my PC. I have issues with its dependencies. Could you please consider using `bcryptjs` npm package instead?

File: server/controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      const user = await userModel.findUserInput(userLogin);
+      if (!user) {
+        return res.status(404).json({
+          status: res.statusCode,
+          error: 'Invalid Login Details'
+        });
+      }
+      const passwordValid = await decryptPassword(password, user.password);
+      if (!passwordValid) {
+        return res.status(401).json({
```
Comment (abdulfatai360) [inline server/controllers/userController.js]: Your HTTP response specification does not follow the team's agreed spec. You may refer to this [https://github.com/omniti-labs/jsend](https://github.com/omniti-labs/jsend) for guidance

File: server/test/root-route.spec.js [hunk 4]
```diff
@@ -5,16 +5,15 @@ import app from '../../index';
 chai.use(chaiHttp);
 const { expect } = chai;
 
-describe('Root URL', () => {
+describe('Root url Test', () => {
   describe('App Root', () => {
-    it('should return status 200 and a message attribute', (done) => {
+    it('should return status 404 and a message attribute', (done) => {
```
Comment (abdulfatai360) [inline server/test/root-route.spec.js]: Bro, well done for the job done so far. But what is the idea behind the `root URL` sending a `404 - Not found` error?

**Review by abdulfatai360 — CHANGES_REQUESTED** [review]
@Xwebyna  Awesome efforts in implementing the signin and signup feature. However, I raised some salient points that needs to be addressed. Please take a look at them.

File: hello-books-swagger.json [hunk 5]
```diff
@@ -1,97 +1,162 @@
 {
```
Comment (tolulope-od) [inline hello-books-swagger.json]: Great work with the controller @Xwebyna I especially like how you've added an error handling middleware to the application. However, could you please let us know why you deleted the old setup of the `swaggerDoc`?

File: server/middlewares/errors.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  *  Error Handling Middleware
+  * @param {object} err
+  * @param {object} req
+  * @param {object} res
+  * @param {function} next
+  * @returns {object} Containing Error Message
+  */
+const errorHandler = (err, req, res, next) => {
+  res.status(500).json({
+    status: res.statusCode,
```
Comment (tolulope-od) [inline server/middlewares/errors.js]: Great work here Ayodeji! I like how you've abstracted the error handling logic. Could you please check however that the response specification is according to what was agreed on for the [JSend](https://github.com/omniti-labs/jsendhttps://github.com/omniti-labs/jsend) specification for server errors? 

Once again, this is fantastic stuff!

File: package.json (hunk 1 above)
Comment (ayodejiAA) [reply package.json]: Thanks. I  might have missed that when resolving conflicts. I will add it in my next commit.

Comment (Xwebyna) [reply package.json]: Thanks. I  might have missed when resolving conflicts. I will add it in my next commit.

File: server/middlewares/errors.js (hunk 6 above)
Comment (ayodejiAA) [reply server/middlewares/errors.js]: Thanks for pointing it out, Tolu.  Necessary adjustments will be made, including updating the swagger doc. Thanks.

File: package.json (hunk 2 above)
Comment (ayodejiAA) [reply package.json]: Okay bro. I will.

File: server/controllers/userController.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+          error: 'Email is already registered'
+        });
+      }
+      data.password = await encryptPassword(password);
+      const user = await userModel.create(data);
+      const token = await generateToken(user);
+
+      return res.status(201).json({
+        status: res.statusCode,
+        token
```
Comment (chialuka) [inline server/controllers/userController.js]: This is massive, @Xwebyna and it is so well done. Thumbs up.

I'd like to know, if all we return is a token on signup, there'll be a need to make a fresh request on the front-end if the user is being redirected tho their dashboard, won't there?

To avoid this situation, won't it be better to return the details of the user except the password?

**Review by chialuka — COMMENTED** [review]
Great job @Xwebyna. This is massive.

**Review by chiboycalix — APPROVED** [review]
Nice job @Xwebyna, your work is well documented in your PR and it's easy to understand. Thumbs up

File: server/controllers/userController.js (hunk 7 above)
Comment (Xwebyna) [reply server/controllers/userController.js]: ThanI thought of that as well

Comment (ayodejiAA) [reply server/controllers/userController.js]: Thanks @LukasChiama. I thought of that as well, but then the story only instructs to implement jwt authentication on the app. So, there will be a fresh request to get the complete profile details of the user when opening the dashboard. Moreover, the frontend will be built with ReactJs which tends to make this easier.

File: hello-books-swagger.json [hunk 8]
```diff
@@ -1,97 +1,200 @@
...
+					"application/json"
+				],
+				"parameters": [
+					{
+						"in": "body",
+						"name": "body",
+						"description": "This is the request body object containing users' sign in information. UserLogin can be email or username",
+						"required": true,
+						"schema": {
+							"$ref": "#/definitions/UserLoginRequest"
```
Comment (tolulope-od) [inline hello-books-swagger.json]: Great work here Ayodeji, thanks for reverting some of the changes on this file. I do have a few concerns here though, I feel that the removal of the User definition doesn't exactly serve the purpose of the documentation. I also don't feel so much confidence about having definitions for login and signup requests. They could just be added as examples in the path method documentation, while definitions would be reserved for our models in the application such as a `User`, `Book` and possibly our `apiResponse` format. Let me know what you think about this.

File: hello-books-swagger.json [hunk 9]
```diff
@@ -1,97 +1,200 @@
...
+		"title": "Hello Books Application API",
+		"contact": {
+			"email": "email@address"
+		},
+		"license": {
+			"name": "Apache 2.0",
+			"url": "http://www.apache.org/licenses/LICENSE-2.0.html"
+		}
+	},
+	"host": "localhost:4000",
```
Comment (tolulope-od) [inline hello-books-swagger.json]: Hey @Xwebyna did you know that setting the host to `localhost:4000` could cause the Documentation to fail when it's hosted on Heroku?

File: server/config/buildTables.js [hunk 10]
```diff
@@ -1,14 +1,18 @@
...
+
 const userTableQuery = `
   CREATE TABLE IF NOT EXISTS users(
     id SERIAL PRIMARY KEY,
-    first_name VARCHAR(100) NOT NULL,
-    last_name VARCHAR(100) NOT NULL,
-    user_name VARCHAR(100) NOT NULL,
+    username VARCHAR(100) UNIQUE NOT NULL,
+    firstname VARCHAR(100) NOT NULL,
+    lastname VARCHAR(100) NOT NULL,
```
Comment (Nerocodes) [inline server/config/buildTables.js]: I'm not sure we decided on a naming convention for our db columns but I think underscores are more readable for serperating words. Are there any particular reasons you chose this?

**Review by Nerocodes — APPROVED** [review]
I left a comment about the naming convention we are using, asides that everything seems good. Great job!

File: server/config/buildTables.js (hunk 10 above)
Comment (ayodejiAA) [reply server/config/buildTables.js]: @Nerocodes Did you see the dB diagram we created? I believe we all agreed on that at the beginning of the sprint.

**Review by tolulope-od — APPROVED** [review]

File: hello-books-swagger.json (hunk 9 above)
Comment (ayodejiAA) [reply hello-books-swagger.json]: Yeah, I added it for development purpose.

**Review by chialuka — APPROVED** [review]

File: hello-books-swagger.json (hunk 8 above)
Comment (ayodejiAA) [reply hello-books-swagger.json]: Thanks for the comment @tolulope-od. I agree with your point. The model should contain the db models not input field definitions.

File: server/config/buildTables.js (hunk 10 above)
Comment (Nerocodes) [reply server/config/buildTables.js]: Alright!

**Review by abdulfatai360 — APPROVED** [review]
Nice work @Xwebyna
